ChatQnA/Dockerfile

```diff
@@ -1,31 +1,51 @@
-
-
 # Copyright (C) 2024 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
 
-FROM python:3.11-slim
+ARG IMAGE_NAME=python
+ARG IMAGE_TAG=3.11-slim
+FROM ${IMAGE_NAME}:${IMAGE_TAG} as base
 
-RUN apt-get update -y && apt-get install -y --no-install-recommends --fix-missing \
+ENV LANG C.UTF-8
+
+RUN apt-get update -y && apt-get install -y \
     libgl1-mesa-glx \
-    libjemalloc-dev \
-    git
+    libjemalloc-dev && \
+    apt-get clean autoclean && \
+    apt-get autoremove -y && \
+    rm -rf /var/lib/apt/lists/*
 
 RUN useradd -m -s /bin/bash user && \
     mkdir -p /home/user && \
     chown -R user /home/user/
 
-WORKDIR /home/user/
-RUN git clone https://github.com/opea-project/GenAIComps.git
-
 WORKDIR /home/user/GenAIComps
+
+FROM base AS devel
+
+ARG GenAICompsRepo="https://github.com/opea-project/GenAIComps"
+ARG GenAICompsBranch="main"  # Branch name or Commit ID
+
+RUN apt-get update -y && apt-get install -y --no-install-recommends --fix-missing \
+    curl
+
+RUN curl -sSL --retry 5 ${GenAICompsRepo}/tarball/${GenAICompsBranch} | tar --strip-components=1 -xzf -
```

Comment:
Using base image from GenAIComps project is the final goal.  These optimization are unnecessary there, and shared base image gives >10x space improvement, when all application images are present.

=> I would skip this and wait for the base image, so that Dockerfiles can be (greatly) simplified to reduce the resulting image size.

Reply:
I compared this against `git clone --depth 1` used in my PR: https://github.com/opea-project/GenAIExamples/pull/1031

Both took about 6 secs (through company VPN), which was a bit disappointing.

I would assume curling compressed tarball to use a bit less bandwidth than Git's "packed" data though.

Extracted tarball took 24MB space. Git clone included extra 14MB in `.git` subdir, but that would not go not go to final target, it's only on the temporary layer.

=> Not compelling enough reason to change my PR at this stage.

Reply:
However, there's a rather large difference in installing `curl` vs. `git` on top of `python:3.11`.

Building curl image took 9s, and git one 18s, probably due to extra dependencies / their sizes (which won't be in the final image):
```
$ docker images |grep  -E "(git|curl)-test"
localhost/git-test                                        latest           d5dfb6922688  2 minutes ago  239 MB
localhost/curl-test                                       latest           0e44588dadfa  3 minutes ago  158 MB
```

=> I'll add note about doing this change to my PR, if I need to update it again.